rstudio/rstudio
# Improve User Experience with XDG directory permissions issues

### Intent

Fixes #4505 

This PR intends to improve the user experience when `~/.local/share/rstudio` and `~/.config/rstudio` are not owned by the correct user. 

### Approach

* On pref save, likely the most common problem for an end user, modify the error message that will be returned to be more descriptive:
![image](https://user-images.githubusercontent.com/37987486/92312050-06589400-ef72-11ea-97f4-c7a9a24e88cb.png)

* In SessionMain, test dirs for write access. If the files exist and the user doesn't have write access, attempt to repair permissions. Log messages on failures:
![image](https://user-images.githubusercontent.com/37987486/92311992-4a976480-ef71-11ea-853e-4453ffc4ca89.png)

RStudio on session start up failure (happens when `~/.local/share/rstudio` is not accessible by the current user):
![image](https://user-images.githubusercontent.com/37987486/92634419-25646800-f289-11ea-8875-0f1302c99510.png)

Downside: Repairing permissions generally fails for the same reason the user cannot write to these files.

### QA Notes

To reproduce this issue, do one of the following:

1. Start RStudio Desktop as root on a system on which RStudio Desktop has never been started.
2. On a system where RStudio Desktop has already been run, execute `chown -R root:root ~/.local/share/rstudio` and/or `chown -R root:root ~/.config/rstudio` to emulate the issue caused by (1).

Then, start RStudio Desktop as your regular user and attempt to modify and save something in your global preferences. Prior to this change you should see an error that says "No such file or directory".

More considerations:
* The session should still start if `~/.config/rstudio` is inaccessible, but it won't start if `~/.local/share/rstudio` is inaccessible.
* This should still work (and the logs should be correct) if the XDG paths are changed. 
* Repairing the permissions (e.g. `chown -R myuser:myuser ~/.config/rstudio`) should resolve the issue completely (i.e. no more logs about it either).
